dart/common/AddonManager.h

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_ADDONMANAGER_H_
+#define DART_COMMON_ADDONMANAGER_H_
+
+#include <map>
+#include <typeinfo>
+#include <typeindex>
+
+#include "dart/common/Addon.h"
+
+namespace dart {
+namespace common {
+
+/// AddonManager is a base class that should be virtually inherited by any class
+/// that wants to be able to manage Addons.
+///
+/// The base AddonManager class is completely agnostic to what kind of Addons it
+/// is given. Addons are stored in a std::map, so access to its Addons happens
+/// on average in log(N) time. Most often, a class that accepts Addons will have
+/// certain Addon types that it will need to access frequently, and it would be
+/// beneficial to have constant-time access to those Addon types. To get
+/// constant-time access to specific Addon types, there are FOUR macros that you
+/// should use in your derived class:
```

Comment:
> Are you currently iterating through the two std::map's in parallel while merging them?

Yep, pretty much that. I iterate between two std::map objects in parallel. If the "receiving" map has an entry that matches the current entry of the "sending" map, then I can use the ``copy`` function to copy the content of the "sender" over to the "receiver". If the "receiving" map does **not** have an entry that matches the current entry of the "sending" map, then I can insert a new (cloned) entry in the **exact** spot on the receiving map where the entry belongs without needing to have ``std::map`` look for the correct spot to place it (or, depending on the semantics of the operation that I'm performing, I can just skip this entry).

The point of this is to be able to save a set of ``Addon::State``s to a map that can then be loaded back in later, or loaded into another copy of a Skeleton. This ordered approach to saving and loading the States allows us to minimize overhead and to be robust to situations where there isn't a perfect match between the types of ``Addon::State``s that you've saved and the types of ``Addon``s present in the Skeleton that you're trying to load the States into.